gradebookng/impl/src/main/java/org/sakaiproject/grading/impl/GradingServiceImpl.java

```diff
@@ -5796,4 +5300,146 @@ private void convertPointsToPercentage(Gradebook gradebook, List<AssignmentGrade
+
+            for (String gbItem : multiSelectorList) {
+                if (StringUtils.isNotBlank(gbItem)) {
+                    String gbUid = getGradebookUidByAssignmentById(siteId, Long.parseLong(gbItem));
+                    gbUidList.add(gbUid);
+                }
+            }
+
+            Collections.sort(gbUidList);
+            Collections.sort(groupList);
+
+            boolean areEqual = gbUidList.equals(groupList);
+
+            if (!areEqual) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public Map<String, String> buildCategoryGradebookMap(List<String> selectedGradebookUids, String categoriesString, String siteId) {
+        Map<String, String> gradebookCategoryMap = new HashMap<>();
+
+        if (categoriesString == null || categoriesString.isBlank()) {
+            selectedGradebookUids.forEach(gbUid -> gradebookCategoryMap.put(gbUid, "-1"));
+        } else {
+            List<String> selectedCategories = Arrays.asList(categoriesString.split(","));
+
+            for (String gbUid : selectedGradebookUids) {
+                List<CategoryDefinition> categoryDefinitions = getCategoryDefinitions(gbUid, siteId);
+
+                String categoryId = categoryDefinitions.stream()
+                    .filter(category -> selectedCategories.contains(category.getId().toString()))
+                    .map(category -> category.getId().toString())
+                    .findFirst()
+                    .orElse("-1");
+
+                    gradebookCategoryMap.put(gbUid, categoryId);
+            }
+        }
+
+        return gradebookCategoryMap;
+    }
+
+    public Long getMatchingUserGradebookItemId(String siteId, String userId, String gradebookItemIdString) {
+        Cache<String, Long> matchingUserGradebookItem = memoryService.getCache(matchingUserGradebookItemCache);
+
+        if (matchingUserGradebookItem != null && matchingUserGradebookItem.containsKey(userId)) {
+            log.debug(buildCacheLogDebug("cacheKeyFound", gradebookGroupInstancesCache));
+            Long gradebookItemId = matchingUserGradebookItem.get(userId);
+
+            if (gradebookItemId != null) {
+                log.debug(buildCacheLogDebug("cacheValueFound", gradebookGroupInstancesCache));
+
+                return gradebookItemId;
+            }
+        }
+
+        List<String> userGradebookList = getGradebookInstancesForUser(siteId, userId);
+        List<String> gradebookItemList = Arrays.asList(gradebookItemIdString.split(","));
+
+        for (String gradebookItem : gradebookItemList) {
+            String foundGradebookUid = getGradebookUidByAssignmentById(siteId,
+                Long.parseLong(gradebookItem));
+
+            if (userGradebookList.contains(foundGradebookUid)) {
+                Long gradebookItemId = Long.valueOf(gradebookItem);
+
+                matchingUserGradebookItem.put(siteId, gradebookItemId);
```

Comment:
This cache appears to check for the userId and then inserts a key using the siteId?

```suggestion
                matchingUserGradebookItem.put(userId, gradebookItemId);
```